src/gsym/parser.rs

```diff
@@ -0,0 +1,424 @@
+        });
+        off += length as usize;
+
+        #[allow(non_upper_case_globals)]
+        match typ {
+            InfoTypeEndOfList => {
+                break;
+            }
+            InfoTypeLineTableInfo | InfoTypeInlineInfo => {}
+            _ => {
+                #[cfg(debug_assertions)]
+                eprintln!("unknown info type");
+            }
+        }
+    }
+
+    data_objs
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::env;
+    use std::fs::File;
+    use std::io::Read;
+    use std::mem;
+    use std::path::Path;
+    use std::ptr::copy;
+
+    #[test]
+    fn test_parse_context() {
+        let args: Vec<String> = env::args().collect();
+        let bin_name = &args[0];
+        let test_gsym = Path::new(bin_name)
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .join("data")
+            .join("test.gsym");
+        let mut gsym_fo = File::open(test_gsym).unwrap();
+        let mut data = vec![];
+
+        gsym_fo.read_to_end(&mut data).unwrap();
+        let ctx = GsymContext::parse_header(&data).unwrap();
+
+        assert_eq!(ctx.addr_at(1), 0x00000000004004b0);
+        let addrinfo = ctx.addr_info(1);
+        assert_eq!(ctx.get_str(addrinfo.name as usize), "_start");
+
+        let idx = 7;
+        assert_eq!(ctx.addr_at(idx), 0x0000000000400596);
+        let addrinfo = ctx.addr_info(idx);
+        assert_eq!(ctx.get_str(addrinfo.name as usize), "factorial");
+
+        let idx = find_address(&ctx, 0x00000000004005db);
+        assert_eq!(idx, 9);
+        let addrinfo = ctx.addr_info(idx);
+        assert_eq!(ctx.get_str(addrinfo.name as usize), "main");
+
+        let addrdata_objs = parse_address_data(addrinfo.data);
+        println!("len = {}", addrdata_objs.len());
+        for o in addrdata_objs {
+            println!("{}", o.typ);
+        }
+    }
+
+    #[test]
+    fn test_find_address() {
+        let args: Vec<String> = env::args().collect();
+        let bin_name = &args[0];
+        let test_gsym = Path::new(bin_name)
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .parent()
+            .unwrap()
+            .join("data")
+            .join("test.gsym");
+        let mut gsym_fo = File::open(test_gsym).unwrap();
+        let mut data = vec![];
+
+        const TEST_SIZE: usize = 6;
+
+        gsym_fo.read_to_end(&mut data).unwrap();
+
+        let mut addr_tab = Vec::<u8>::new();
+        addr_tab.resize(TEST_SIZE * 4, 0);
+
+        let mut ctx = GsymContext::parse_header(&data).unwrap();
+        ctx.header.num_addrs = TEST_SIZE as u32;
+        ctx.header.addr_off_size = 4;
+        ctx.header.base_address = 0;
+        // Share the same buffer.
+        ctx.addr_tab = unsafe { mem::transmute(addr_tab.as_slice()) };
```

Comment:
Safety comment is missing and this operation is not safe at all. Rust does not allow the presence of shared reference while a mutable one is outstanding and you create a mutable one below.

Reply:
Got it

Reply:
I just wrote a safety comment in the test case.  Please check it.

Reply:
Unfortunately not. It does not matter if both execute at the same time or not. It's not about data races, it's about aliasing. Can you instantiate a new test case fixture for each iteration, perhaps?

Reply:
I am not sure what you expect in safety comments. Rust doesn't like to have mutable and immutable references to the same object at the same time. The problem is not aliasing.  The problem is how we use references.  Here, the comment is to address the potential problem to explain why it is safe.

By the way, I will create a new instance of `ctx` every time to avoid this.